### PR TITLE
Keep pruning cross-validation folds stable across epochs

### LINKED ISSUE #317: Pruning with shuffled CV leaks training samples into validation across epochs
With pruning enabled, `OptunaSearchCV` calls `cv.split()` again on every training epoch while retaining the same fitted estimators. A shuffled splitter backed by a `RandomState` instance yields different folds each time, so an estimator is evaluated on samples it previously trained on.

Reproduced on main `f1ba03843b2af96fe6b584aa9583c8dc776b5a80`, Python 3.10/Linux. This scorer checks the held-out contract while using real `SGDClassifier.partial_fit`:

```python
import numpy as np
from sklearn.linear_model import SGDClassifier
from sklearn.model_selection import KFold
from optuna_integration import OptunaSearchCV

class TrackingClassifier(SGDClassifier):
    def partial_fit(self, X, y, classes=None):
        super().partial_fit(X, y, classes=classes)
        self.seen_samples_ = getattr(self, "seen_samples_", set()) | set(X[:, 0])
        return self

def score(estimator, X, y):
    assert estimator.seen_samples_.isdisjoint(X[:, 0])
    return estimator.score(X, y)

search = OptunaSearchCV(
    TrackingClassifier(random_state=0), {},
    cv=KFold(3, shuffle=True, random_state=np.random.RandomState(0)),
    enable_pruning=True, max_iter=3, n_trials=1,
    error_score="raise", scoring=score, refit=False,
)
search.fit(np.arange(30).reshape(-1, 1), np.arange(30) % 2)
```

The assertion fails on the second epoch. Expected: each estimator keeps its training and validation split for the duration of a trial. AI assistance was used for investigation and local verification.

### PR DESCRIPTION
**Pull request by betacatsling**
## Motivation

Pruning keeps a fitted estimator for each CV fold across epochs, but recreates the folds on every epoch. A shuffled splitter with a mutable `RandomState` then puts previously trained samples into the validation fold, contaminating its score.

## Description of the changes

Materialize the splits once per trial and reuse them for its epochs. This preserves per-trial splitter behavior while keeping each estimator's held-out samples fixed.

The regression trains actual `SGDClassifier` instances and asserts that validation samples never appear in their accumulated training samples, across three epochs and two trials. It fails on the unchanged base. Local sklearn integration suite: **30 passed**. Changed-file Black, isort, flake8 and mypy checks passed. Other integration suites were not run.

AI assistance was used for implementation and local verification.

Fixes #317.

### CODE REVIEW

**Review by not522 — APPROVED** [review]
LGTM!
